<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Space X  Falcon 9 First Stage Landing Prediction**


## Assignment:  Machine Learning Prediction


Estimated time needed: **60** minutes


Space X advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against space X for a rocket launch.   In this lab, you will create a machine learning pipeline  to predict if the first stage will land given the data from the preceding labs.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/api/Images/landing_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/api/Images/crash.gif)


Most unsuccessful landings are planed. Space X; performs a controlled landing in the oceans.


## Objectives


Perform exploratory  Data Analysis and determine Training Labels

*   create a column for the class
*   Standardize the data
*   Split into training data and test data

\-Find best Hyperparameter for SVM, Classification Trees and Logistic Regression

*   Find the method performs best using test data


## Import Libraries and Define Auxiliary Functions


In [1]:
import piplite
await piplite.install(['numpy'])
await piplite.install(['pandas'])
await piplite.install(['seaborn'])

We will import the following libraries for the lab


In [2]:
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np
# Matplotlib is a plotting library for python and pyplot gives us a MatLab like plotting framework. We will use this in our plotter function to plot data.
import matplotlib.pyplot as plt
#Seaborn is a Python data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics
import seaborn as sns
# Preprocessing allows us to standarsize our data
from sklearn import preprocessing
# Allows us to split our data into training and testing data
from sklearn.model_selection import train_test_split
# Allows us to test parameters of classification algorithms and find the best one
from sklearn.model_selection import GridSearchCV
# Logistic Regression classification algorithm
from sklearn.linear_model import LogisticRegression
# Support Vector Machine classification algorithm
from sklearn.svm import SVC
# Decision Tree classification algorithm
from sklearn.tree import DecisionTreeClassifier
# K Nearest Neighbors classification algorithm
from sklearn.neighbors import KNeighborsClassifier

This function is to plot the confusion matrix.


In [3]:
def plot_confusion_matrix(y,y_predict):
    "this function plots the confusion matrix"
    from sklearn.metrics import confusion_matrix

    cm = confusion_matrix(y, y_predict)
    ax= plt.subplot()
    sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
    ax.set_xlabel('Predicted labels')
    ax.set_ylabel('True labels')
    ax.set_title('Confusion Matrix'); 
    ax.xaxis.set_ticklabels(['did not land', 'land']); ax.yaxis.set_ticklabels(['did not land', 'landed']) 
    plt.show() 

## Load the dataframe


Load the data


In [4]:
from js import fetch
import io

URL1 = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/dataset_part_2.csv"
resp1 = await fetch(URL1)
text1 = io.BytesIO((await resp1.arrayBuffer()).to_py())
data = pd.read_csv(text1)

In [5]:
data.head()

FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1  2010-06-04       Falcon 9  6104.959412   LEO  CCAFS SLC 40   
1             2  2012-05-22       Falcon 9   525.000000   LEO  CCAFS SLC 40   
2             3  2013-03-01       Falcon 9   677.000000   ISS  CCAFS SLC 40   
3             4  2013-09-29       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5  2013-12-03       Falcon 9  3170.000000   GTO  CCAFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False        NaN    1.0   
1    None None        1     False   False  False        NaN    1.0   
2    None None        1     False   False  False        NaN    1.0   
3  False Ocean        1     False   False  False        NaN    1.0   
4    None None        1     False   False  False        NaN    1.0   

   ReusedCount Serial   Longitude   Latitude  Class  
0            0  B0003  -80.577366  28.561857      0  
1            0  B0005  -80.577366  28.561857      0  
2            0  B0007  -80.577366  28.561857      0  
3            0  B1003 -120.610829  34.632093      0  
4            0  B1004  -80.577366  28.561857      0

In [6]:
URL2 = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/dataset_part_3.csv'
resp2 = await fetch(URL2)
text2 = io.BytesIO((await resp2.arrayBuffer()).to_py())
X = pd.read_csv(text2)

In [7]:
X.head(100)

FlightNumber   PayloadMass  Flights  Block  ReusedCount  Orbit_ES-L1  \
0            1.0   6104.959412      1.0    1.0          0.0          0.0   
1            2.0    525.000000      1.0    1.0          0.0          0.0   
2            3.0    677.000000      1.0    1.0          0.0          0.0   
3            4.0    500.000000      1.0    1.0          0.0          0.0   
4            5.0   3170.000000      1.0    1.0          0.0          0.0   
..           ...           ...      ...    ...          ...          ...   
85          86.0  15400.000000      2.0    5.0          2.0          0.0   
86          87.0  15400.000000      3.0    5.0          2.0          0.0   
87          88.0  15400.000000      6.0    5.0          5.0          0.0   
88          89.0  15400.000000      3.0    5.0          2.0          0.0   
89          90.0   3681.000000      1.0    5.0          0.0          0.0   

    Orbit_GEO  Orbit_GTO  Orbit_HEO  Orbit_ISS  ...  Serial_B1058  \
0         0.0        0.0        0.0        0.0  ...           0.0   
1         0.0        0.0        0.0        0.0  ...           0.0   
2         0.0        0.0        0.0        1.0  ...           0.0   
3         0.0        0.0        0.0        0.0  ...           0.0   
4         0.0        1.0        0.0        0.0  ...           0.0   
..        ...        ...        ...        ...  ...           ...   
85        0.0        0.0        0.0        0.0  ...           0.0   
86        0.0        0.0        0.0        0.0  ...           1.0   
87        0.0        0.0        0.0        0.0  ...           0.0   
88        0.0        0.0        0.0        0.0  ...           0.0   
89        0.0        0.0        0.0        0.0  ...           0.0   

    Serial_B1059  Serial_B1060  Serial_B1062  GridFins_False  GridFins_True  \
0            0.0           0.0           0.0             1.0            0.0   
1            0.0           0.0           0.0             1.0            0.0   
2            0.0           0.0           0.0             1.0            0.0   
3            0.0           0.0           0.0             1.0            0.0   
4            0.0           0.0           0.0             1.0            0.0   
..           ...           ...           ...             ...            ...   
85           0.0           1.0           0.0             0.0            1.0   
86           0.0           0.0           0.0             0.0            1.0   
87           0.0           0.0           0.0             0.0            1.0   
88           0.0           1.0           0.0             0.0            1.0   
89           0.0           0.0           1.0             0.0            1.0   

    Reused_False  Reused_True  Legs_False  Legs_True  
0            1.0          0.0         1.0        0.0  
1            1.0          0.0         1.0        0.0  
2            1.0          0.0         1.0        0.0  
3            1.0          0.0         1.0        0.0  
4            1.0          0.0         1.0        0.0  
..           ...          ...         ...        ...  
85           0.0          1.0         0.0        1.0  
86           0.0          1.0         0.0        1.0  
87           0.0          1.0         0.0        1.0  
88           0.0          1.0         0.0        1.0  
89           1.0          0.0         0.0        1.0  

[90 rows x 83 columns]

## TASK  1


Create a NumPy array from the column <code>Class</code> in <code>data</code>, by applying the method <code>to_numpy()</code>  then
assign it  to the variable <code>Y</code>,make sure the output is a  Pandas series (only one bracket df\['name of  column']).


In [8]:
Y = data['Class'].to_numpy()

In [9]:
Y

array([0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1], dtype=int64)

## TASK  2


Standardize the data in <code>X</code> then reassign it to the variable  <code>X</code> using the transform provided below.


In [10]:
# students get this 
transform = preprocessing.StandardScaler()

In [11]:
X_OLD = X
X_OLD

FlightNumber   PayloadMass  Flights  Block  ReusedCount  Orbit_ES-L1  \
0            1.0   6104.959412      1.0    1.0          0.0          0.0   
1            2.0    525.000000      1.0    1.0          0.0          0.0   
2            3.0    677.000000      1.0    1.0          0.0          0.0   
3            4.0    500.000000      1.0    1.0          0.0          0.0   
4            5.0   3170.000000      1.0    1.0          0.0          0.0   
..           ...           ...      ...    ...          ...          ...   
85          86.0  15400.000000      2.0    5.0          2.0          0.0   
86          87.0  15400.000000      3.0    5.0          2.0          0.0   
87          88.0  15400.000000      6.0    5.0          5.0          0.0   
88          89.0  15400.000000      3.0    5.0          2.0          0.0   
89          90.0   3681.000000      1.0    5.0          0.0          0.0   

    Orbit_GEO  Orbit_GTO  Orbit_HEO  Orbit_ISS  ...  Serial_B1058  \
0         0.0        0.0        0.0        0.0  ...           0.0   
1         0.0        0.0        0.0        0.0  ...           0.0   
2         0.0        0.0        0.0        1.0  ...           0.0   
3         0.0        0.0        0.0        0.0  ...           0.0   
4         0.0        1.0        0.0        0.0  ...           0.0   
..        ...        ...        ...        ...  ...           ...   
85        0.0        0.0        0.0        0.0  ...           0.0   
86        0.0        0.0        0.0        0.0  ...           1.0   
87        0.0        0.0        0.0        0.0  ...           0.0   
88        0.0        0.0        0.0        0.0  ...           0.0   
89        0.0        0.0        0.0        0.0  ...           0.0   

    Serial_B1059  Serial_B1060  Serial_B1062  GridFins_False  GridFins_True  \
0            0.0           0.0           0.0             1.0            0.0   
1            0.0           0.0           0.0             1.0            0.0   
2            0.0           0.0           0.0             1.0            0.0   
3            0.0           0.0           0.0             1.0            0.0   
4            0.0           0.0           0.0             1.0            0.0   
..           ...           ...           ...             ...            ...   
85           0.0           1.0           0.0             0.0            1.0   
86           0.0           0.0           0.0             0.0            1.0   
87           0.0           0.0           0.0             0.0            1.0   
88           0.0           1.0           0.0             0.0            1.0   
89           0.0           0.0           1.0             0.0            1.0   

    Reused_False  Reused_True  Legs_False  Legs_True  
0            1.0          0.0         1.0        0.0  
1            1.0          0.0         1.0        0.0  
2            1.0          0.0         1.0        0.0  
3            1.0          0.0         1.0        0.0  
4            1.0          0.0         1.0        0.0  
..           ...          ...         ...        ...  
85           0.0          1.0         0.0        1.0  
86           0.0          1.0         0.0        1.0  
87           0.0          1.0         0.0        1.0  
88           0.0          1.0         0.0        1.0  
89           1.0          0.0         0.0        1.0  

[90 rows x 83 columns]

In [12]:
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]

array([[-1.71291154e+00, -1.94814463e-16, -6.53912840e-01,
        -1.57589457e+00, -9.73440458e-01, -1.05999788e-01,
        -1.05999788e-01, -6.54653671e-01, -1.05999788e-01,
        -5.51677284e-01,  3.44342023e+00, -1.85695338e-01,
        -3.33333333e-01, -1.05999788e-01, -2.42535625e-01,
        -4.29197538e-01,  7.97724035e-01, -5.68796459e-01,
        -4.10890702e-01, -4.10890702e-01, -1.50755672e-01,
        -7.97724035e-01, -1.50755672e-01, -3.92232270e-01,
         9.43398113e+00, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.50755672e-01, -1.05999788e-01,
        -1.05999788e-01, -1.05999788e-01, -1.05999788e-01,
        -1.05999788e-01, -1.50755672e-01, -1.05999788e-0

We split the data into training and testing data using the  function  <code>train_test_split</code>.   The training data is divided into validation data, a second set used for training  data; then the models are trained and hyperparameters are selected using the function <code>GridSearchCV</code>.


## TASK  3


Use the function train_test_split to split the data X and Y into training and test data. Set the parameter test_size to  0.2 and random_state to 2. The training data and test data should be assigned to the following labels.


<code>X_train, X_test, Y_train, Y_test</code>


In [13]:
from sklearn.model_selection import train_test_split
#X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size=0.2, random_state=4)
X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size=0.2, random_state=2)
print ('Train set:', X_train.shape,  Y_train.shape)
print ('Test set:', X_test.shape,  Y_test.shape)

Train set: (72, 83) (72,)
Test set: (18, 83) (18,)


In [14]:
X_train
X_test
Y_train
Y_test

array([1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1], dtype=int64)

we can see we only have 18 test samples.


In [15]:
Y_test.shape

(18,)

## TASK  4


Create a logistic regression object  then create a  GridSearchCV object  <code>logreg_cv</code> with cv = 10.  Fit the object to find the best parameters from the dictionary <code>parameters</code>.


In [16]:
parameters ={'C':[0.01,0.1,1],
             'penalty':['l2'],
             'solver':['lbfgs']}

In [17]:
parameters ={"C":[0.01,0.1,1],'penalty':['l2'], 'solver':['lbfgs']}# l1 lasso l2 ridge
lr=LogisticRegression()


In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C=0.01, solver='liblinear').fit(X_train,Y_train)
LR
logreg_cv = LogisticRegression(C=10, solver='liblinear').fit(X_train,Y_train)
logreg_cv

LogisticRegression(C=10, solver='liblinear')

We output the <code>GridSearchCV</code> object for logistic regression. We display the best parameters using the data attribute <code>best_params\_</code> and the accuracy on the validation data using the data attribute <code>best_score\_</code>.


In [19]:
print("tuned hpyerparameters :(best parameters) ",logreg_cv.best_params_)
print("accuracy :",logreg_cv.best_score_)

<class 'AttributeError'>: 'LogisticRegression' object has no attribute 'best_params_'

In [20]:
Yhat = LR.predict(X_test)
Yhat

array([0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0], dtype=int64)

In [21]:
Yhat_prob = LR.predict_proba(X_test)
Yhat_prob

array([[0.51137002, 0.48862998],
       [0.30504064, 0.69495936],
       [0.39317677, 0.60682323],
       [0.30408613, 0.69591387],
       [0.44007676, 0.55992324],
       [0.30702126, 0.69297874],
       [0.33829549, 0.66170451],
       [0.78972443, 0.21027557],
       [0.31569153, 0.68430847],
       [0.754274  , 0.245726  ],
       [0.76513396, 0.23486604],
       [0.5573471 , 0.4426529 ],
       [0.50656148, 0.49343852],
       [0.30355875, 0.69644125],
       [0.40811952, 0.59188048],
       [0.40957211, 0.59042789],
       [0.55773771, 0.44226229],
       [0.51487308, 0.48512692]])

## TASK  5


Calculate the accuracy on the test data using the method <code>score</code>:


In [22]:
from sklearn.metrics import jaccard_score
jaccard_score(Y_test, Yhat,pos_label=0)

0.5555555555555556

Lets look at the confusion matrix:


In [23]:
from sklearn.metrics import classification_report, confusion_matrix
import itertools
def plot_confusion_matrix(cm, classes,
                          normalize=False,
                          title='Confusion matrix',
                          cmap=plt.cm.Blues):
    """
    This function prints and plots the confusion matrix.
    Normalization can be applied by setting `normalize=True`.
    """
    if normalize:
        cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
        print("Normalized confusion matrix")
    else:
        print('Confusion matrix, without normalization')

    print(cm)

    plt.imshow(cm, interpolation='nearest', cmap=cmap)
    plt.title(title)
    plt.colorbar()
    tick_marks = np.arange(len(classes))
    plt.xticks(tick_marks, classes, rotation=45)
    plt.yticks(tick_marks, classes)

    fmt = '.2f' if normalize else 'd'
    thresh = cm.max() / 2.
    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
        plt.text(j, i, format(cm[i, j], fmt),
                 horizontalalignment="center",
                 color="white" if cm[i, j] > thresh else "black")

    plt.tight_layout()
    plt.ylabel('True label')
    plt.xlabel('Predicted label')
print(confusion_matrix(Y_test, Yhat, labels=[1,0]))

[[9 3]
 [1 5]]


In [24]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[9 3]
 [1 5]]


In [25]:
yhat=logreg_cv.predict(X_test)
plot_confusion_matrix(Y_test,yhat)

Confusion matrix, without normalization
[1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1 0 1]


<class 'TypeError'>: Invalid shape (18,) for image data

Examining the confusion matrix, we see that logistic regression can distinguish between the different classes.  We see that the major problem is false positives.


## TASK  6


Create a support vector machine object then  create a  <code>GridSearchCV</code> object  <code>svm_cv</code> with cv - 10.  Fit the object to find the best parameters from the dictionary <code>parameters</code>.


In [26]:
parameters = {'kernel':('linear', 'rbf','poly','rbf', 'sigmoid'),
              'C': np.logspace(-3, 3, 5),
              'gamma':np.logspace(-3, 3, 5)}
svm = SVC()

We tries several kernel which are linear, rbf, and sigmoid
The best performer is sigmoid

In [27]:
from sklearn import svm
clf = svm.SVC(kernel='linear')
clf.fit(X_train, Y_train)

SVC(kernel='linear')

In [28]:
Yhat = clf.predict(X_test)
Yhat [0:5]

array([1, 0, 1, 1, 1], dtype=int64)

In [29]:
Yhat

array([1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1], dtype=int64)

In [30]:
Y_test

array([1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1], dtype=int64)

In [31]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[9 3]
 [3 3]]


In [32]:
print("tuned hpyerparameters :(best parameters) ",svm_cv.best_params_)
print("accuracy :",svm_cv.best_score_)

<class 'NameError'>: name 'svm_cv' is not defined

## TASK  7


Calculate the accuracy on the test data using the method <code>score</code>:


In [33]:
from sklearn.metrics import f1_score
f1_score(Y_test, Yhat, average='weighted') 

0.6666666666666666

We can plot the confusion matrix


In [34]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[9 3]
 [3 3]]


Those for linear
Thus we repeat for task 6 and 7 for rbf and sigmoid

In [35]:
yhat=svm_cv.predict(X_test)
plot_confusion_matrix(Y_test,yhat)

<class 'NameError'>: name 'svm_cv' is not defined

RBF begin here

In [36]:
from sklearn import svm
clf = svm.SVC(kernel='rbf')
clf.fit(X_train, Y_train)

SVC()

In [37]:
Yhat = clf.predict(X_test)
Yhat [0:5]

array([0, 1, 1, 1, 1], dtype=int64)

In [38]:
from sklearn.metrics import f1_score
f1_score(Y_test, Yhat, average='weighted') 

0.764102564102564

In [39]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[11  1]
 [ 3  3]]


Sigmoid here

In [40]:
from sklearn import svm
clf = svm.SVC(kernel='sigmoid')
clf.fit(X_train, Y_train)

SVC(kernel='sigmoid')

In [41]:
Yhat = clf.predict(X_test)
Yhat [0:5]

array([1, 1, 1, 1, 1], dtype=int64)

In [42]:
from sklearn.metrics import f1_score
f1_score(Y_test, Yhat, average='weighted') 

0.8148148148148149

In [43]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[12  0]
 [ 3  3]]


## TASK  8


Create a decision tree classifier object then  create a  <code>GridSearchCV</code> object  <code>tree_cv</code> with cv = 10.  Fit the object to find the best parameters from the dictionary <code>parameters</code>.


In [44]:
parameters = {'criterion': ['gini', 'entropy'],
     'splitter': ['best', 'random'],
     'max_depth': [2*n for n in range(1,10)],
     'max_features': ['auto', 'sqrt'],
     'min_samples_leaf': [1, 2, 4],
     'min_samples_split': [2, 5, 10]}

tree = DecisionTreeClassifier()

In [45]:
drugTree = DecisionTreeClassifier(criterion="entropy", max_depth = 10)
drugTree # it shows the default parameters

DecisionTreeClassifier(criterion='entropy', max_depth=10)

In [46]:
drugTree.fit(X_train,Y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=10)

In [47]:
predTree = drugTree.predict(X_test)

In [48]:
print (predTree [0:5])
print (Y_test [0:5])

[1 1 1 1 1]
[1 1 1 1 1]


In [49]:
from sklearn import metrics
import matplotlib.pyplot as plt
print("DecisionTrees's Accuracy: ", metrics.accuracy_score(Y_test, predTree))

DecisionTrees's Accuracy:  0.8888888888888888


In [50]:
tree.plot_tree(drugTree)
plt.show()

<class 'AttributeError'>: 'DecisionTreeClassifier' object has no attribute 'plot_tree'

In [51]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[12  0]
 [ 3  3]]


In [52]:
print("tuned hpyerparameters :(best parameters) ",tree_cv.best_params_)
print("accuracy :",tree_cv.best_score_)

<class 'NameError'>: name 'tree_cv' is not defined

## TASK  9


Calculate the accuracy of tree_cv on the test data using the method <code>score</code>:


In [53]:
from sklearn import metrics
import matplotlib.pyplot as plt
print("DecisionTrees's Accuracy: ", metrics.accuracy_score(Y_test, predTree))

DecisionTrees's Accuracy:  0.8888888888888888


We can plot the confusion matrix


In [54]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[12  0]
 [ 3  3]]


In [55]:
yhat = svm_cv.predict(X_test)
plot_confusion_matrix(Y_test,yhat)

<class 'NameError'>: name 'svm_cv' is not defined

## TASK  10


Create a k nearest neighbors object then  create a  <code>GridSearchCV</code> object  <code>knn_cv</code> with cv = 10.  Fit the object to find the best parameters from the dictionary <code>parameters</code>.


In [56]:
parameters = {'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
              'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
              'p': [1,2]}

KNN = KNeighborsClassifier()

In [57]:
from sklearn.neighbors import KNeighborsClassifier

In [58]:
k = 10
#Train Model and Predict  
neigh = KNeighborsClassifier(n_neighbors = k).fit(X_train,Y_train)
neigh

KNeighborsClassifier(n_neighbors=10)

In [59]:
Yhat = neigh.predict(X_test)
Yhat[0:5]

array([1, 1, 1, 1, 1], dtype=int64)

In [60]:
from sklearn import metrics
print("Train set Accuracy: ", metrics.accuracy_score(Y_train, neigh.predict(X_train)))
print("Test set Accuracy: ", metrics.accuracy_score(Y_test, Yhat))

Train set Accuracy:  0.6666666666666666
Test set Accuracy:  0.6666666666666666


In [21]:
print("tuned hpyerparameters :(best parameters) ",knn_cv.best_params_)
print("accuracy :",knn_cv.best_score_)

<class 'NameError'>: name 'knn_cv' is not defined

## TASK  11


Calculate the accuracy of knn_cv on the test data using the method <code>score</code>:


In [61]:
from sklearn import metrics
print("Train set Accuracy: ", metrics.accuracy_score(Y_train, neigh.predict(X_train)))
print("Test set Accuracy: ", metrics.accuracy_score(Y_test, Yhat))

Train set Accuracy:  0.6666666666666666
Test set Accuracy:  0.6666666666666666


We can plot the confusion matrix


In [62]:
# Compute confusion matrix
cnf_matrix = confusion_matrix(Y_test, Yhat, labels=[1,0])
np.set_printoptions(precision=2)


# Plot non-normalized confusion matrix
plt.figure()
plot_confusion_matrix(cnf_matrix, classes=['churn=1','churn=0'],normalize= False,  title='Confusion matrix')

Confusion matrix, without normalization
[[12  0]
 [ 6  0]]


In [22]:
yhat = knn_cv.predict(X_test)
plot_confusion_matrix(Y_test,yhat)

<class 'NameError'>: name 'knn_cv' is not defined

## TASK  12


Find the method performs best:


We list the acurracy of each model
Log linear 0.56
svm linear 0.67
svm rbf 0.76
svm sigmoid 0.81
decision tree 0.89
KNN 0.67
Thus the best model is decision tree

## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01)


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
